RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -0,0 +1,891 @@
+          } // end if deltaR
+        } // end of loop over patatracks
+
+        // normalize to sum and define stdDev
+        for (int eta_idx = 0; eta_idx<nCellEta ; eta_idx++){
+          for (int phi_idx = 0; phi_idx<nCellPhi ; phi_idx++){
+            if(getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum)>0.){
+              getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDeltaEta) = getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDeltaEta) / getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum);
+              getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDeltaPhi) = getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDeltaPhi) / getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum);
+              getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackChi2OverNdof) = getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackChi2OverNdof) / getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum);
+              getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackNdof) = getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackNdof) / getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum);
+              getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDxy) = getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDxy) / getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum);
+              getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDz) = getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackDz) / getCell(tau_idx, phi_idx, eta_idx, NNInputs::PatatrackPtSum);
+            } // end if
+          } // end loop on phi
+        } // end loop on eta
+      } // end loop on taus
+   } // end function
+
+
+   std::vector<float> L2TauNNProducer::GetTauScore(const tensorflow::Tensor &cellGridMatrix){
+     std::vector<tensorflow::Tensor> pred_tensor;
+     std::vector<float> pred_vector;
+     tensorflow::run(L2cacheData->session, {{inputTensorName_, cellGridMatrix}}, {outputTensorName_}, &pred_tensor);
+     std::vector<int> tensor_shape;
+     for(int d=0; d<cellGridMatrix.shape().dims(); d++) {
+         tensor_shape.push_back(cellGridMatrix.shape().dim_size(d));
+     }
+     for(int tau_idx=0; tau_idx < tensor_shape.at(0); tau_idx++){
+       pred_vector.push_back(pred_tensor[0].matrix<float>()(tau_idx, 0));
+     }
+
+     return pred_vector;
+   }
+
+
+   void L2TauNNProducer::produce(edm::Event& event, const edm::EventSetup& eventsetup){
+
+     std::map<size_t, l1t::TauVectorRef> TauCollectionMap;   // map of l1Taucollections - input_index
+     std::vector<std::string> CollectionNamesMap;
+     CollectionNamesMap.reserve(L1TauDesc.size());
+     l1t::TauVectorRef allTaus;
+
+     for (size_t inp_idx = 0; inp_idx<L1TauDesc.size(); inp_idx++){
+       CollectionNamesMap.push_back(L1TauDesc[inp_idx].CollectionName) ;
+
+       edm::Handle<trigger::TriggerFilterObjectWithRefs> l1TriggeredTaus;
+       event.getByToken(L1TauDesc[inp_idx].input_token, l1TriggeredTaus);
```

Comment:
here and below define a utility function to retrieve handle from the event
```c++
template<typename T>
edm::Handle<T> getHandle(const edm::Event& event, const edm::EDGetTokenT<T>& token)
{
  edm::Handle<T> handle;
  event.getByToken(token, handle);
  return handle;
}
```
and then use it like
```c++
const auto l1TriggeredTaus = getHandle(event, L1TauDesc[inp_idx].input_token);
```